pkg/kademlia/kademlia.go

```diff
@@ -5,250 +5,199 @@ package kademlia
 
 import (
 	"context"
-	"crypto/rand"
-	"log"
+	"errors"
+	"fmt"
 	"net"
-	"strconv"
 
-	bkad "github.com/coyle/kademlia"
 	"github.com/zeebo/errs"
+	"google.golang.org/grpc"
 
 	"storj.io/storj/pkg/dht"
+	"storj.io/storj/pkg/node"
+	"storj.io/storj/pkg/provider"
 	proto "storj.io/storj/protos/overlay"
 )
 
+const (
+	alpha                       = 5
+	defaultIDLength             = 256
+	defaultBucketSize           = 20
+	defaultReplacementCacheSize = 5
+)
+
 // NodeErr is the class for all errors pertaining to node operations
 var NodeErr = errs.Class("node error")
 
+// BootstrapErr is the class for all errors pertaining to bootstrapping a node
+var BootstrapErr = errs.Class("bootstrap node error")
+
 //TODO: shouldn't default to TCP but not sure what to do yet
 var defaultTransport = proto.NodeTransport_TCP
 
+// NodeNotFound is returned when a lookup can not produce the requested node
+var NodeNotFound = NodeErr.New("node not found")
+
+type lookupOpts struct {
+	amount int
+}
+
 // Kademlia is an implementation of kademlia adhering to the DHT interface.
 type Kademlia struct {
-	routingTable   RoutingTable
+	alpha          int // alpha is a system wide concurrency parameter
+	routingTable   *RoutingTable
 	bootstrapNodes []proto.Node
-	ip             string
-	port           string
+	address        string
 	stun           bool
-	dht            *bkad.DHT
+	nodeClient     node.Client
+	identity       *provider.FullIdentity
 }
 
 // NewKademlia returns a newly configured Kademlia instance
-func NewKademlia(id dht.NodeID, bootstrapNodes []proto.Node, ip string, port string) (*Kademlia, error) {
-	if port == "" {
-		return nil, NodeErr.New("must specify port in request to NewKademlia")
-	}
-
-	ips, err := net.LookupIP(ip)
+func NewKademlia(id dht.NodeID, bootstrapNodes []proto.Node, address string, identity *provider.FullIdentity) (*Kademlia, error) {
```

Comment:
good point. I'll create another ticket to do that since it will be another change that requires a lot of other changes to tests and throughout the code base.

Reply:
https://storjlabs.atlassian.net/browse/V3-429